sdks/java/core/src/test/java/org/apache/beam/sdk/util/ReshuffleTest.java

```diff
@@ -88,6 +94,33 @@ public void testJustReshuffle() {
     pipeline.run();
   }
 
+  /**
+   * Tests that timestamps can
+   */
+  @Test
+  @Category(RunnableOnService.class)
+  public void testReshuffleAndAssignTimestamps() {
+    PCollection<KV<String, Integer>> input = pipeline
+        .apply(Create.of(ARBITRARY_KVS)
+            .withCoder(KvCoder.of(StringUtf8Coder.of(), VarIntCoder.of())));
+
+    PCollection<KV<String, Integer>> output =
+        input
+            .apply(Reshuffle.<String, Integer>of())
+            .apply(
+                ParDo.of(
+                    new DoFn<KV<String, Integer>, KV<String, Integer>>() {
+                      @ProcessElement
+                      public void assertMinValueTimestamp(ProcessContext ctxt) {
+                        assertThat(ctxt.timestamp(), equalTo(BoundedWindow.TIMESTAMP_MIN_VALUE));
```

Comment:
The issue is that the Reshuffle can still reassign timestamps, but only backwards in time (to the earliest element in the pane). Generally that's less unsafe than moving elements forwards in time, because the runner can do so safely but once it moves them forwards it's dangerous for users to move them back.

I've restructured the test to assert the proper relationship between the timestamps.